fix(ivy): adding event listeners for global objects (window, document, body)

AndrewKushnir opened the pull request:
This update introduces support for global object (window, document, body) listeners, that can be defined via host listeners on Components and Directives.

PR resolves issue FW-743.

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
fix(ivy): adding event listeners for global objects (window, document, body)

This update introduces support for global object (window, document, body) listeners, that can be defined via host listeners on Components and Directives.

mary-poppins commented:
You can preview 92d0b2d at https://pr27772-92d0b2d.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/src/render3/util.ts:
```diff
@@ -101,6 +101,30 @@ export function getNativeByTNode(tNode: TNode, hostView: LView): RElement|RText|
   return readElementValue(hostView[tNode.index]);
 }
 
+export function extractEventListenerDetails(eventName: string, element: RElement):
```
Would it be possible to have this splitting logic in the compiler instead and modify (or better yet - introduce a new) listener instruction? My concern is that this adds more code / processing on runtime. On top of this the code is not tree-shakable and will be pulled in even if there are no global events listened to in the application.

pkozlowski-opensource reviewed (changes requested):
I would love to discuss the possibility of moving more logic to the compiler side and thus making the "global events" logic tree-shakable

AndrewKushnir commented:
Thanks for the feedback @pkozlowski-opensource! I will have a look at this today and let you know.

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

mary-poppins commented:
You can preview 991994d at https://pr27772-991994d.ngbuilds.io/.

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

JoostK commented on packages/core/src/render3/interfaces/renderer.ts:
```diff
@@ -26,6 +26,8 @@ export enum RendererStyleFlags3 {
 
 export type Renderer3 = ObjectOrientedRenderer3 | ProceduralRenderer3;
 
+export type GlobalTargetSelector = 'document' | 'window' | 'body';
```
I find the name to be confusing, as I interpret "selector" as CSS related which these strings are not. Would `GlobalTargetName` or `GlobalTargetIdentifier` be an option?

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -792,10 +792,16 @@ export function locateHostElement(
 export function listener(
-    eventName: string, listenerFn: (e?: any) => any, useCapture = false): void {
+    eventName: string, listenerFn: (e?: any) => any, useCapture = false,
+    globalTargetGetter?: (element: any) => {
```
Could we name this `targetResolver` or something? To me, `globalTargetGetter` is odd.

pkozlowski-opensource replied:
+1 maybe `eventTargetResolver`?

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -807,6 +813,8 @@ export function listener(
   // add native event listener - applicable to elements only
   if (tNode.type === TNodeType.Element) {
     const native = getNativeByTNode(tNode, lView) as RElement;
+    const glob = globalTargetGetter ? globalTargetGetter(native) : {} as any;
```
`glob` usually means the file globbing tool :)

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -816,15 +824,22 @@ export function listener(
+      // The first argument of `listen` function in Procedural Renderer is:
+      // - either a target name (as a string) in case of global target (window, document, body)
+      // - or element reference (in all other cases)
+      const cleanupFn = renderer.listen(glob.selector || target, eventName, listenerFn);
```
Why pass only the "selector" here? Wouldn't it be nice if the procedural renderer had access to the resolver fn as well as the name (because it seems the name is only necessary for `Renderer2` compat).

AndrewKushnir replied:
Thanks for the comment @JoostK. I'll keep it as is to support Render2 contract. Changing existing behavior of Renderer2 is outside of the scope of this task.

pkozlowski-opensource commented on packages/core/src/render3/util.ts:
```diff
@@ -101,6 +101,24 @@ export function getNativeByTNode(tNode: TNode, hostView: LView): RElement|RText|
+    }
+    if (selector === 'window') {
+      return wrap(element.ownerDocument.defaultView);
+    }
```
Yep, this would be my preferred approach as well

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

mary-poppins commented:
You can preview 3ac2eac at https://pr27772-3ac2eac.ngbuilds.io/.

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

mary-poppins commented:
You can preview a65120c at https://pr27772-a65120c.ngbuilds.io/.

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

mary-poppins commented:
You can preview c8c3ad8 at https://pr27772-c8c3ad8.ngbuilds.io/.

AndrewKushnir commented:
Thanks for the review everyone!

I've created 3 fixup commits (to address different parts of the feedback):

- removing duplicate logic in event params generation in compiler: https://github.com/angular/angular/pull/27772/commits/3ac2eac552f5ee70aa167708df23f2b9aba88f7d
- introduce specific function to resolve window/document/body: https://github.com/angular/angular/pull/27772/commits/a65120cf20678dd14e60646cb8dfed09a20ce1d6
- extract global target tests into separate functions: https://github.com/angular/angular/pull/27772/commits/c8c3ad8208bdb145de87f8e873f96031c32ec7c7

Please have a look again and let me know if you have any additional feedback/comments.

Thank you.

pkozlowski-opensource reviewed (approved):
This LGTM to now, thnx for making all those changes @AndrewKushnir !

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

mary-poppins commented:
You can preview 2c31ed7 at https://pr27772-2c31ed7.ngbuilds.io/.

mary-poppins commented:
You can preview b83ef5d at https://pr27772-b83ef5d.ngbuilds.io/.

AndrewKushnir commented:
Thanks for the review @pkozlowski-opensource!

FYI, I've added extra test (in [a separate fixup commit](https://github.com/angular/angular/pull/27772/commits/b83ef5d468f311c5e5d5e923f2ce9596675d5cbc)) to check whether we unsubscribe correctly, as Kara suggested. In order to cleanup (remove global listeners) once the test is done, I've introduced simple `destroy` method for `ComponentFixture` and `TemplateFixture` to invoke all LView cleanup machinery (we should probably call this method for each Fixture after each test?).

@kara could you please have a look at this commit: https://github.com/angular/angular/pull/27772/commits/b83ef5d468f311c5e5d5e923f2ce9596675d5cbc and let me know if you have any feedback?

P.S. I also rebased to the latest master to resolve conflicts.

Thank you.

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

mary-poppins commented:
You can preview dbc5885 at https://pr27772-dbc5885.ngbuilds.io/.

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

mary-poppins commented:
You can preview 66ce7d6 at https://pr27772-66ce7d6.ngbuilds.io/.

trotyl commented on packages/core/src/render3/util.ts:
```diff
@@ -288,3 +288,15 @@ export function findComponentView(lView: LView, declarationMode?: boolean): LVie
 }
+
+export function resolveWindow(element: RElement & {ownerDocument: Document}) {
+  return {name: 'window', target: element.ownerDocument.defaultView};
```
Isn't it breaking custom `Renderer`? The API requirement on `Renderer`-site is only `RElement`:

https://github.com/angular/angular/blob/f8096d499324cf0961f092944bbaedd05364eea1/packages/core/src/render3/interfaces/renderer.ts#L61

So that a conforming `Renderer` implementation could throw `cannot read property 'defaultView' of undefined` here, right?

AndrewKushnir committed:
fixup! fix(ivy): adding event listeners for global objects (window, document, body)

mary-poppins commented:
You can preview 581fec4 at https://pr27772-581fec4.ngbuilds.io/.

kara reviewed (approved):
LGTM

kara commented:
[presubmit](http://test/OCL:228248486:BASE:228248674:1546904044109:ea86d3ac)
